api/v1alpha4/cluster_webhook_test.go

```diff
@@ -46,45 +46,197 @@ func TestClusterDefault(t *testing.T) {
 	g.Expect(c.Spec.ControlPlaneRef.Namespace).To(Equal(c.Namespace))
 }
 
-func TestClusterValidation(t *testing.T) {
-	valid := &Cluster{
+func TestClusterDefaultTopologyVersion(t *testing.T) {
+	g := NewWithT(t)
+
+	c := &Cluster{
 		ObjectMeta: metav1.ObjectMeta{
-			Namespace: "foo",
+			Namespace: "fooboo",
 		},
 		Spec: ClusterSpec{
-			ControlPlaneRef: &corev1.ObjectReference{
-				Namespace: "foo",
-			},
-			InfrastructureRef: &corev1.ObjectReference{
-				Namespace: "foo",
+			Topology: &Topology{
+				Class:   "foo",
+				Version: "1.19.1",
 			},
 		},
 	}
-	invalidInfraNamespace := valid.DeepCopy()
-	invalidInfraNamespace.Spec.InfrastructureRef.Namespace = "bar"
 
-	invalidCPNamespace := valid.DeepCopy()
-	invalidCPNamespace.Spec.InfrastructureRef.Namespace = "baz"
+	t.Run("for Cluster", utildefaulting.DefaultValidateTest(c))
+	c.Default()
 
+	g.Expect(c.Spec.Topology.Version).To(HavePrefix("v"))
+}
+
+func TestClusterValidation(t *testing.T) {
 	tests := []struct {
 		name      string
+		in        *Cluster
+		old       *Cluster
 		expectErr bool
-		c         *Cluster
 	}{
 		{
 			name:      "should return error when cluster namespace and infrastructure ref namespace mismatch",
 			expectErr: true,
-			c:         invalidInfraNamespace,
+			in: &Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "foo",
+				},
+				Spec: ClusterSpec{
+					InfrastructureRef: &corev1.ObjectReference{
+						Namespace: "bar",
+					},
+					ControlPlaneRef: &corev1.ObjectReference{
+						Namespace: "foo",
+					},
+				},
+			},
 		},
 		{
 			name:      "should return error when cluster namespace and controlplane ref namespace mismatch",
 			expectErr: true,
-			c:         invalidCPNamespace,
+			in: &Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "foo",
+				},
+				Spec: ClusterSpec{
+					InfrastructureRef: &corev1.ObjectReference{
+						Namespace: "foo",
+					},
+					ControlPlaneRef: &corev1.ObjectReference{
+						Namespace: "bar",
+					},
+				},
+			},
 		},
 		{
 			name:      "should succeed when namespaces match",
 			expectErr: false,
-			c:         valid,
+			in: &Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "foo",
+				},
+				Spec: ClusterSpec{
+					ControlPlaneRef: &corev1.ObjectReference{
+						Namespace: "foo",
+					},
+					InfrastructureRef: &corev1.ObjectReference{
+						Namespace: "foo",
+					},
+				},
+			},
+		},
+		{
+			name:      "should return error when topology does not have class",
+			expectErr: true,
+			in: &Cluster{
+				Spec: ClusterSpec{
+					Topology: &Topology{},
+				},
+			},
+		},
+		{
+			name:      "should return error when topology does not have valid version",
+			expectErr: true,
+			in: &Cluster{
+				Spec: ClusterSpec{
+					Topology: &Topology{
+						Class:   "foo",
+						Version: "invalid",
+					},
+				},
+			},
+		},
+		{
+			name:      "should return error when duplicated MachineDeployments names exists in a Topology",
```

Comment:
I think a version of this test which returns successfully would be great (basically copy&paste and change one of the deployment names)